### DISCUSSION TITLE
Running NomadNet on Windows - the index.mu does not seem to want to load

### DISCUSSION THREAD

**Discussion by KenSamson** [markqvist/Reticulum#636] [Nomad Network]
![image](https://github.com/user-attachments/assets/cd3a17e3-6412-4014-b0f9-6e6c53456922)

I've worked for the last few weeks on and off to get a windows hosted NomadNet  dirt simple index.mu to load from a Meshchat client.

Should be simple as can be, right?     Tried everything I could think of, a dozen times.

So... I'm running out of options.     So I go to my raspberry pi, throw the same index.mu on it, enable the node, make sure an LXMF propagation node is set as default...   go back to Meshchat on my windows laptop, and bingo, it comes up big as life.  (all two lines of it)

Does anyone else out there host NomadNet with pages under windows 11, and what is it that I could possibly be doing wrong?

I don't really think it's a bug...  but at this point it confuses me.

RadioManAlpha

**Comment by liamcottle** [comment]
I've just investigated this, as I thought it might be a MeshChat bug.

But I'm seeing `request failed` when hosting nomadnet pages from a Windows machine, even when using nomadnet as a client on another machine.

**Nomadnet Node (Hosting the Site)**
- Windows 10
- nomadnet v0.5.6
- serving default index.mu

**Nomadnet Client (Browsing the Site)**
- MacBook Pro
- nomadnet v0.5.3,
- nomadnet v0.5.6 (upgraded to this version, same issue)

I do note that the Windows machine that is hosting the node, that nothing else seems to be able to browse, can in fact browse other nomadnet nodes fine, such as the test node on the MacBook.

Look like it'll be some sort of bug with nomadnet on Windows.

**Comment by KenSamson** [comment]
![image](https://github.com/user-attachments/assets/46931ea2-eb0e-457f-b8c5-fa907d0b021f)

It seems like NomadNet is trying to call an application in windows that is not a valid Win32 application.  

If that's true, then others won't be able to host micron pages on a windows box running nomadnet either...   can anyone else confirm if they are doing this?

**Reply by liamcottle** [reply]
Yep, same log on my end too.

**Reply by liamcottle** [reply]
Looks like it's trying to execute the file.

https://github.com/markqvist/NomadNet/blob/d8cfc69ac63faf462f95411efd07d8351b2061a5/nomadnet/Node.py#L176

**Reply by liamcottle** [reply]
Pinging @markqvist. Any thoughts on this one?

I don't think Windows supports executing arbitrary files, like Linux/Mac does when putting the `#!/bin/env python` at the top of the file etc.

Maybe for now, if serving from Windows, we just jump to serving plain text?

**Reply by liamcottle** [reply]
Downloading files by manually pasting in a nomadnet file URL into MeshChat works as expected, so it's just getting caught up on the fact that the _page_ files are marked as executable on Windows (doesn't look possible to remove that), so it tries to execute the page file, which results in the `not a valid Win32 application`.

**Reply by markqvist** [reply]
> Maybe for now, if serving from Windows, we just jump to serving plain text?

That's probably the right way to go about it. On NIXes, it will check for file execute permission, and if it is an executable, will run it - if not, just send the contents.

If windows doesn't have any such mechanism, it's probably best to just leave off that functionality on Windows.

**Reply by KenSamson** [reply]
Can anyone confirm that they are able to serve up even plain text on windows?

I changed my index.mu to one line text, no  # in front.. and it still never loads.    It could be my path config.. so if other people say they can serve pages on windows, it would be great to know what you have your pages directory set to..


If nobody can do it, then it looks like we can't serve pages at all from a windows host.

**Reply by liamcottle** [reply]
Negative. It's because the code checks if the file is executable, which on Windows, it appears all files are. I couldn't even remove that attribute from any files even as administrator... Probably need to put in a PR to check if we are running on Windows, if so, skip trying to execute and just always return plain text.

**Reply by liamcottle** [reply]
I've opened a PR to fix this: https://github.com/markqvist/NomadNet/pull/65
